### ISSUE TITLE
[FR/enhancement]: Synchronize highlights across separate KOReader devices

### ISSUE THREAD

**Issue by arooni** [koreader/koreader#4587]
* KOReader version: 2019.02.01 stable version
* Device: Kindle PW4 / Google Pixel XL / Kindle Fire 10 Tablet

#### Issue
So seeing as I do all my reading on KOreader on my PW now, I figure I might as well take advantage of progress sync/whispersync (works great!!!) so I can read on any of my devices and always have my progress synced.

I tend to highlight quite a few passages, and then export those to Evernote (also, amazing feature).  But it seems that highlights don't sync across devices.  

This to me means if I made some highlights on separate devices, I'd need to export the notes to Evernote each time.. so potentially for one book I'd need to export the highlighted passages three times (if I were reading on my PW, tablet/phone).  I'm also unsure of what happens when there's an existing evernote note for a given book.  I'd assume the most recent device to export to evernote would overwrite any prior notes.  

I don't know how many other people read across multiple devices, but this would seem to be a very useful feature, and would hopefully lead to more people running koreader in more places and on more devices.

NOTE: I did check other issues to see if this had been suggested before: #628, but didn't see anything.  If I'm remiss and this is already available someone please correct me.  Thanks!

#### Steps to reproduce

**Comment by Frenzie** [comment]
@embryo10 made a KOReader Highlights utility that can now also do syncing.
https://www.mobileread.com/forums/showpost.php?p=3816259&postcount=40

**Comment by arooni** [comment]
Does Issue #4670 address this?

Looked into that utility but apparently it only runs on Windows and I use Ubuntu + Mac :\.  

It's cool that syncing highlights is coming along... definitely one fo the features I'm wishing for the most.

**Comment by noembryo** [comment]
It's also running on Ubuntu (Python 2.7.x) as script.
I just made it to run on Python 3 too.
Just have to upload it and update the dependencies..

**Comment by Frenzie** [comment]
@arooni You'll have to install the dependencies (PySide and future; Python will be installed by default) and then it should run. :-)

**Comment by noembryo** [comment]
>But bookmark with highlight sync is not available for multiple devices using KOHighlight or any other plugin. Is someone working on it or has a plan to implement it?

Highlight sync for multiple devices _is_ available in KOHighlights, but manually, not automatically.
I also can't use online sync, since that will probably need some server/subscription and extra networking knowledge that I don't currently have.

**Comment by uncharted-user** [comment]
> Highlight sync for multiple devices _is_ available in KOHighlights, but manually, not automatically. I also can't use online sync, since that will probably need some server/subscription and extra networking knowledge that I don't currently have.

Yes, I used that but what I was looking something like,

- Make highlight + bookmark on Android and export.
- Import exported highlight + bookmark from Android into other device like Kindle or Linux App Image.

I am currently manually duplicating highlight line by line from one device to another, which is really annoying (3+ devices) to do, and I miss highlights or notes quite often.

**Comment by noembryo** [comment]
> I am currently manually duplicating highlight line by line from one device to another, which is really annoying (3+ devices) to do, and I miss highlights or notes quite often.

I don't understand why you prefer  that, over: 
* Make highlight (and nothing more on device)
* Connect device to KOHighlights > "Archive" highlights
* Connect other device with KOHighlights > "Sync with archived"
* Connect other device with KOHighlights > "Sync with archived"
* ...

It is true that I don't use it like this, but the option is there.

**Comment by noembryo** [comment]
Also, if you can connect all your devices at once, you can use the option "Sync with file", and use your most recent `metadata.xxx.lua` file to sync the others..

**Comment by uncharted-user** [comment]
Unfortunately none of the methods worked. What I found lua file highlighted notes are synced. But those highlights are not being shown on pages where it supposed to be.
 
I also saw this warning during the process,
![Screenshot from 2022-10-21 23-32-18](https://user-images.githubusercontent.com/100132960/197263632-b1260bf5-dcbd-407e-a4f5-bda78671ff71.png)

**Comment by noembryo** [comment]
>Unfortunately none of the methods worked. What I found lua file highlighted notes are synced. But those highlights are not being shown on pages where it supposed to be.

Well, they should have worked.
If not, please open an issue at [KOHighlights](https://github.com/noembryo/KoHighlights)'s repository, so we don't pollute this issue.
You have to be a lot more specific though.. :wink: 

>I also saw this warning during the process,

This is a normal warning that you can accept.
It has to do with the original page number of the highlights.
The highlights themselves should be identical though, and at the right place.

**Comment by uncharted-user** [comment]
But, I observed the following,

- All the highlight + notes + bookmarks synced if I manually copied the sdr folder along with epub from Android and pasted in anywhere on Linux (this is probably KOHighlight sync with the file does which also worked)and vice versa.

I don't know whether it is supposed to be for all epub or there are exceptions. This will also work for me if works on Kindle devices too. 

Another problem was KOHighlight browser does not see my mounted phone or windows drive!

**Comment by noembryo** [comment]
> this is probably KOHighlight sync with the file does which also worked

No, this is not it, but please, let's move this to [KOHighlights](https://github.com/noembryo/KoHighlights)'s repository..

**Comment by moshin34** [comment]
Is there a technical limitation here?  I find it quite odd that KOreader doesnt have a native wireless sync for highlights/notes.   Especial for someting that was requested 5-6 years ago.   Just curious if this is something that will never, ever happen.  I do not like the idea of plugging in, or installing syncthing, but if I have to go KLUGY I will. 

I appreciate all the effort people put into KOreader, it has really chnaged my life for the better. 

Thanks!

**Comment by Frenzie** [comment]
Anything that happens will have to grab someone's interest. Since that hasn't happened yet I'd say chances aren't good that'll change in the near future, but you never know.

**Comment by uncharted-user** [comment]
> I believe SyncThing has a way to perform some custom operation after synchronization. A host could merge highlights after each sync and then sync the result to other peers.
> 
> However, it sems KOHighlight is for interactive usage only, so I don't know if it is actually feasible or not
Could you please provide us with a reference or a tutorial?

**Comment by hius07** [comment]
Not as automized as position syncing, but may be useful.
https://github.com/koreader/koreader/pull/13372

**Comment by uncharted-user** [comment]
> Already in the nightly build: https://build.koreader.rocks/download/nightly/v2024.11-230-gcfbb676d7_2025-03-10/

Thank you so much.

**Comment by pirofti** [comment]
@hius07, that is great work indeed! Would you be willing to do that for highlights as well? Anyway we can help?
